src/CakeSlayer/Error.hs

```diff
@@ -0,0 +1,101 @@
+{- | This module contains types ans functions for the app error customization.
+-}
+
+module CakeSlayer.Error
+       ( -- * Pure errors handling
+         WithError
+       , ErrorWithSource (..)
+       , throwError
+
+         -- * Exceptions
+       , AppException (..)
+       , toNoSourceException
+
+         -- * Helper unctions
+       , throwOnNothing
+       , throwOnNothingM
+
+         -- * 'SourcePosition' helpers
+       , SourcePosition (..)
+       , toSourcePosition
+       ) where
+
+import Control.Monad.Except (MonadError)
+import GHC.Stack (SrcLoc (SrcLoc, srcLocModule, srcLocStartLine))
+
+import qualified Control.Monad.Except as E (throwError)
+
+
+{- | Type alias for errors that have acess to 'CallStack'. Specicialise this
+constraint in your code to monomorphic error data type like this:
+
+@
+__type__ WithError = CakeSlayer.'WithError' MyError
+@
+-}
+type WithError err m = (MonadError (ErrorWithSource err) m, HasCallStack)
+
+{- | Wrapper around error type with attached source code position.
+-}
+data ErrorWithSource err = ErrorWithSource
+    { appErrorCallStack :: !SourcePosition
+    , appErrorType      :: !err
+    } deriving stock (Show, Eq, Functor)
+
+{- | Specialized version of 'E.throwError' that attaches source code position of
+the place where this error was thrown.
+-}
+throwError :: WithError err m => err -> m a
+throwError = E.throwError . ErrorWithSource (toSourcePosition callStack)
+{-# INLINE throwError #-}
+
+newtype SourcePosition = SourcePosition Text
```

Comment:
We usually using named fields for newtypes, I guess we should apply this rule here too

Reply:
And also let's add haddock comment as we exporting this.